File: turing_machine/turing_machine.py
from typing import Dict
from turing_machine.constants import LAMBDA, STOP_STATE, MOVE_LEFT, MOVE_NONE, MOVE_RIGHT
from turing_machine.constants import NORMAL_MODE, BY_STEP_MODE
from turing_machine.constants import SUCCESSFUL_STATUS, MAX_ITERATIONS_REACHED_STATUS, MAX_ITERATIONS
from turing_machine.tape import Tape


class TuringMachine:
    """
    Turing machine class.

    :param alphabet: the alphabet of this machine
    :type alphabet: string of characters
    :param rules: maps state, character to [symbol, move, next state]
    :type rules: {str: {str: [str]}}
    """
    def __init__(self, *, alphabet: str, rules: Dict[str, Dict[str, list]], tape: str = '', position: int = 0):
        self.alphabet = alphabet + LAMBDA
        self.rules = rules
        self.tape = Tape(tape)
        self.position = position

    def __print_line(self):
        """prints horisonatal line of the rules tabel"""
        print('+--------' * len(self.alphabet) + '+--------+')

    def __print_header(self):
        """prints header of the rules tabel"""
        print("|        |", " | ".join("%6s" % c for c in self.alphabet), "|")

    def __cell_to_str(self, q: str, c: str, prettify: bool) -> str:
        """returns string to print in corresponding cell"""

        c_next, move, q_next = self.rules[q][c]

        if prettify and c_next == c:
            if q_next == q:
                return move

            if move == MOVE_NONE:
                return q_next

        return c_next + " " + q_next + " " + move

    def __print_state(self, q: str, prettify: bool):
        """Prints whole rules tabel line corresponding to the state
        :param q: the state
        """
        cells = " | ".join("%6s" % self.__cell_to_str(q, c, prettify) for c in self.alphabet)
        print("| %6s |" % q, cells, "|")

    def print_rules(self, prettify_rules: bool = True):
        """Prints rules as plain text table.

        :param prettify_rules: whether to simplify (char,move,state)-cell when possible
        """
        print("Alphabet:", self.alphabet)
        print("Rules table:")

        self.__print_line()
        self.__print_header()
        self.__print_line()

        for q in self.rules:
            self.__print_state(q, prettify_rules)

        self.__print_line()
        print()

    def print_tape(self, with_position=True):
        """Prints current state of the tape.

        :param with_position: whether to mark current head position with brackets
        """
        tape_str = self.tape.string_with_position(self.position) if with_position else str(self.tape)
        print(tape_str or "Tape is empty")

    def set_tape_string(self, tape: str, position: int = 0):
        """Set up tape and head position

        :param tape: what is written on the new tape
        :param position: where the head is on the new tape
        """
        self.position = position
        self.tape = Tape(tape)

    def get_tape_string(self):
        """Returns the current state of the tape as a string"""
        return str(self.tape)

    def run(self, mode: str = NORMAL_MODE, max_tacts: int = MAX_ITERATIONS, initial_state: str = "q0") -> dict:
        """Emulate the Turing machine.

        :param mode: whether to include result of every step in return
        :returns: dictionary with fields:
            * status: whether the machine stoped by itself (successfully) or because of tacts limit
            * result: what is written on the tape as result
            * iterations: how many tacts it run
            * head_position: where the head is on the tape
            * steps: list of intemideate information for every step, included only for "by step" mode
        """
        q = initial_state
        tacts = 0
        steps = []

        while q != STOP_STATE and tacts < max_tacts:
            c = self.tape[self.position]
            c_next, move, q_next = self.rules[q][c]
            self.tape[self.position] = c_next

            if mode == BY_STEP_MODE:
                steps.append({
                    "curr_state": q,
                    "next_state": q_next,
                    "curr_character": c,
                    "next_character": c_next,
                    "move": move,
                    "tact": tacts
                })

            if move == MOVE_RIGHT:
                self.position += 1
            elif move == MOVE_LEFT:
                self.position -= 1

            tacts += 1
            q = q_next

        result = {
            "status": SUCCESSFUL_STATUS if tacts < max_tacts else MAX_ITERATIONS_REACHED_STATUS,
            "result": self.get_tape_string(),
            "iterations": tacts,
            "head_position": self.position
        }

        if mode == BY_STEP_MODE:
            result["steps"] = steps

        return result

File: turing_machine/web/js/turing_machine.js
function TuringMachine(alphabet = 'abc', tape = 'abacaba', rules = []) {
    this.tapeBox = document.getElementById('tape')
    this.alphabetBox = document.getElementById('alphabet')
    this.rulesBox = document.getElementById('rules')
    this.resultBox = document.getElementById('result-box')
    this.alphabet = alphabet + 'λ'

    this.InitAlphabet(alphabet)
    this.InitTape(tape)
    this.InitRules(rules)
}

TuringMachine.prototype.InitTape = function(tape) {
    for (let i = 0; i < tape.length; i++) {
        let box = document.createElement('div')
        box.innerHTML = tape[i]
        box.className = 'tape-box'
        this.tapeBox.appendChild(box)
    }
}

TuringMachine.prototype.InitAlphabet = function(alphabet) {
    for (let i = 0; i < alphabet.length; i++) {
        let box = document.createElement('div')
        box.innerHTML = alphabet[i]
        box.className = 'tape-box'
        this.alphabetBox.appendChild(box)
    }
}

TuringMachine.prototype.InitRules = function(rules) {
    states = Object.keys(rules)

    let header = document.createElement('tr')

    for (let i = 0; i < this.alphabet.length + 1; i++) {
        let cell = document.createElement('th')

        if (i == 0) {
            cell.innerHTML = 'q \\ a'
        }
        else {
            cell.innerHTML = this.alphabet[i - 1]
        }

        header.appendChild(cell)
    }

    this.rulesBox.appendChild(header)

    for (let state of states) {
        let row = document.createElement('tr')

        keys = Object.keys(rules[state])

        let cell = document.createElement('td')
        cell.innerHTML = state
        row.appendChild(cell)

        for (let i = 0; i < keys.length; i++) {
            let cell = document.createElement('td')
            cell.innerHTML = rules[state][keys[i]].join(' ')
            row.appendChild(cell)
        }

        this.rulesBox.appendChild(row)
    }
}

TuringMachine.prototype.AddRowWithValues = function(table, values, elem='td') {
    let row = document.createElement('tr')

    for (let i = 0; i < values.length; i++) {
        let cell = document.createElement(elem)
        cell.innerHTML = values[i]
        row.appendChild(cell)
    }

    table.appendChild(row)
}

TuringMachine.prototype.Run = function(result) {
    this.resultBox.innerHTML = ''
    this.resultBox.innerHTML += '<b>Result: </b>' + result["result"] + '<br>'
    this.resultBox.innerHTML += '<b>Iterations: </b>' + result["iterations"] + '<br>'
    this.resultBox.innerHTML += '<b>Status: </b>' + result["status"] + '<br>'
    this.resultBox.innerHTML += '<b>Head position: </b>' + result["head_position"] + '<br><br>'

    this.resultBox.innerHTML += '<b>Steps:</b><br>'

    let table = document.createElement('table')
    this.AddRowWithValues(table, ['curr state', 'curr char', 'next state', 'next char', 'move'], 'th')

    for (let i = 0; i < result["steps"].length; i++) {
        let step = result["steps"][i]
        let curr_state = step["curr_state"]
        let next_state = step["next_state"]
        let curr_char = step["curr_character"]
        let next_char = step["next_character"]
        let move = step["move"]

        this.AddRowWithValues(table, [curr_state, curr_char, next_state, next_char, move])
    }

    this.resultBox.appendChild(table)
}
